identity/src/main/java/com/stripe/android/identity/states/ScanState.kt

```diff
@@ -0,0 +1,132 @@
+                )
+                Found(type)
+            } else {
+                Log.d(
+                    TAG,
+                    "Model outputs ${analyzerOutput.category}, which doesn't match with " +
+                        "scanType $type, stay in Initial"
+                )
+                this
+            }
+
+        private fun Category.matchesScanType(scanType: ScanType): Boolean {
+            return this == Category.ID_BACK && scanType == ScanType.ID_BACK ||
+                this == Category.ID_FRONT && scanType == ScanType.ID_FRONT ||
+                this == Category.PASSPORT && scanType == ScanType.PASSPORT
+        }
+    }
+
+    /**
+     * State when scan has found the required type, the machine could stay in this state for a
+     * while if more image needs to be processed to reach the next state.
+     */
+    internal class Found(type: ScanType) : ScanState(type) {
+        override fun consumeTransition(analyzerOutput: AnalyzerOutput) =
+            when {
+                isUnsatisfied() -> {
+                    val reason = "unsatisfied reason"
+                    Log.d(
+                        TAG, "Satisfaction check fails due to $reason, transition to Unsatisfied."
+                    )
+                    Unsatisfied(reason, type)
+                }
+                moreResultsRequired() -> {
+                    Log.d(TAG, "More results needed, stay in Found.")
+                    this
+                }
+                else -> {
+                    Log.d(TAG, "Satisfaction check succeeds, transition to Satisfied.")
+                    Satisfied(type)
+                }
+            }
+
+        /**
+         * Determine if more images should be processed before reaching [Satisfied].
+         *
+         * TODO(ccen) - Introduce conditions that requires more results
+         */
+        private fun moreResultsRequired(): Boolean {
+            return false
+        }
+
+        /**
+         * Determine if satisfaction failed and should transition to [Unsatisfied].
+         *
+         * TODO(ccen) - Introduce unsatisfied reasons
+         */
+        private fun isUnsatisfied(): Boolean {
+            return false
+        }
+    }
+
+    /**
+     * State  when satisfaction checking passed.
+     */
+    internal class Satisfied(type: ScanType) : ScanState(type) {
+        override fun consumeTransition(analyzerOutput: AnalyzerOutput): ScanState {
+            Log.d(TAG, "Scan for $type Satisfied, transition to Finished.")
+            return Finished(type)
+        }
+    }
+
+    /**
+     * State  when satisfaction checking failed.
+     */
+    internal class Unsatisfied(private val reason: String, type: ScanType) : ScanState(type) {
+        override fun consumeTransition(analyzerOutput: AnalyzerOutput): ScanState {
+            Log.d(TAG, "Scan for $type Unsatisfied with reason $reason, transition to Initial.")
+            return Initial(type)
+        }
+    }
```

Comment:
What is the purpose of these states? It looks like they simply consume a single analyzer result and always transition to another state. Is there a reason not to just go from Found -> Finished, or Found -> Initial?

Reply:
These two states indicate some signals for certain UI, please refer to [this](https://paper.dropbox.com/doc/State-machine-for-Identity-Scan--BbgXvpLLubT18kpsgJUju071Ag-Qs7xK4pTkkqxWlgCaLyTj#:uid=199060037651666807984634&h2=The-states) table for some info. 

`Satisified` would show user that a capture is successful, then transitioning to `Finished` would show the frozen image with a check mar.
`Unsatisfied` would briefly signal the user something like "we lost the focus of your camera", before transitioning back to `Initial` where it shows "position the doc in the center"

I guess we could also somehow achieve that by adding some `val`s inside the `Found` instance, but I feel like this is a cleaner way. wdty?

Reply:
This is definitely the cleaner way to do that. The timing on these transitions would be dictated by the output of the analyzer. I.E. the first analyzer result received while in these state will transition to a "finished" state.

Are you OK relying on the timing of the analyzer, or should there be some timer within the state itself? something like:

```
    internal class Satisfied(type: ScanType) : ScanState(type) {
        override fun consumeTransition(analyzerOutput: AnalyzerOutput): ScanState {
            if (reachedStateAt.elapsedSince() > DISPLAY_SATISFIED_DURATION) {
                Log.d(TAG, "Scan for $type Satisfied, transition to Finished.")
                return Finished(type)
            } else {
                Log.d(TAG, "Displaying satisfied state, waiting for timeout")
                return this
            }
        }
    }
```

Reply:
👍 added the `ClockMark` from `camera-core`, also added a test.
The actual timeout value might be updated in the future